### PR TITLE
tests(driver): type check

### PR DESCRIPTION
**Pull request by connorjclark**
Previous work: #6874

### CODE REVIEW

File: lighthouse-core/test/gather/driver-test.js [hunk 1]
```diff
@@ -40,20 +40,28 @@ function makePromiseInspectable(promise) {
     throw err;
   });
 
-  inspectablePromise.isDone = () => isResolved || isRejected;
-  inspectablePromise.isResolved = () => isResolved;
-  inspectablePromise.isRejected = () => isRejected;
-  inspectablePromise.getDebugValues = () => ({resolvedValue, rejectionError});
-
-  return inspectablePromise;
+  return Object.assign(inspectablePromise, {
```
Comment (connorjclark) [inline lighthouse-core/test/gather/driver-test.js]: This pattern allows type script to infer the type correctly, while not changing the prototype of the return value (it is still a Promise).

File: lighthouse-core/test/gather/mock-commands.js [hunk 2]
```diff
@@ -33,20 +33,37 @@ function createMockSendCommandFn() {
...
-
-  mockFn.findInvocation = (command, sessionId) => {
-    expect(mockFn).toHaveBeenCalledWith(command, sessionId, expect.anything());
-    return mockFn.mock.calls.find(call => call[0] === command && call[1] === sessionId)[2];
-  };
+  const mockFn = Object.assign(mockFnImpl, {
+    /**
+     * @template {keyof LH.CrdpCommands} C
+     * @param {C} command
+     * @param {RecursivePartial<LH.CrdpCommands[C]['returnType']>=} response
```
Comment (connorjclark) [inline lighthouse-core/test/gather/mock-commands.js]: Test code often doesn't provide a full protocol response, so this makes everything optional (yet well-typed for the properties that are given).

File: types/externs.d.ts [hunk 3]
```diff
@@ -28,8 +28,9 @@ declare global {
 
   /** Make optional all properties on T and any properties on object properties of T. */
   type RecursivePartial<T> = {
-    [P in keyof T]+?: T[P] extends object ?
-      RecursivePartial<T[P]> :
+    [P in keyof T]+?:
```
Comment (connorjclark) [inline types/externs.d.ts]: This change adds array support.

File: lighthouse-core/test/gather/driver-test.js [hunk 4]
```diff
@@ -81,14 +89,19 @@ async function flushAllTimersAndMicrotasks() {
   }
 }
 
+/** @type {Driver & {on: ReturnType<typeof createMockOnceFn>, once: ReturnType<typeof createMockOnceFn>}} */
 let driver;
+/** @type {Connection & {sendCommand: ReturnType<typeof createMockSendCommandFn>}} */
 let connectionStub;
 
 beforeEach(() => {
+  // @ts-ignore
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]:
```suggestion
  // @ts-ignore - connectionStub has a mocked version of sendCommand implemented in each test
```

File: lighthouse-core/test/gather/driver-test.js [hunk 5]
```diff
@@ -81,14 +89,19 @@ async function flushAllTimersAndMicrotasks() {
...
   connectionStub = new Connection();
+  // @ts-ignore
   connectionStub.sendCommand = cmd => {
     throw new Error(`${cmd} not implemented`);
   };
+  // @ts-ignore
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]:
```suggestion
  // @ts-ignore - driver has a mocked version of on/once  implemented in each test
```

File: lighthouse-core/test/gather/driver-test.js [hunk 6]
```diff
@@ -431,29 +442,34 @@ describe('.gotoURL', () => {
...
+    class ReplayConnection extends Connection {
       connect() {
         return Promise.resolve();
       }
       disconnect() {
         return Promise.resolve();
       }
       replayLog() {
-        redirectDevtoolsLog.forEach(msg => this.emit('protocolevent', msg));
+        // @ts-ignore
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]: why do we need this one isn't redirectDevtoolsLog any? ts-ignore explanation would be nice :)

File: lighthouse-core/test/gather/driver-test.js [hunk 7]
```diff
@@ -587,6 +604,7 @@ describe('.gotoURL', () => {
       driver._waitForNetworkIdle = createMockWaitForFn();
       driver._waitForCPUIdle = createMockWaitForFn();
 
+      // @ts-ignore
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]: is this so we don't need to create an entire passcontext?

File: lighthouse-core/test/gather/driver-test.js [hunk 8]
```diff
@@ -524,6 +540,7 @@ describe('.gotoURL', () => {
         driver._waitForNetworkIdle = createMockWaitForFn();
         driver._waitForCPUIdle = createMockWaitForFn();
 
+        // @ts-ignore
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]:
```suggestion
        // @ts-ignore - dynamic property access, tests will definitely fail if the property were to change
```

File: lighthouse-core/test/gather/driver-test.js [hunk 9]
```diff
@@ -876,24 +905,27 @@ describe('Domain.enable/disable State', () => {
...
+    await driver.sendCommand('Network.enable');
+    await driver.sendCommand('Network.enable');
     expect(connectionStub.sendCommand).toHaveBeenCalledTimes(1);
 
-    await driver.sendCommand('Network.disable', {});
+    await driver.sendCommand('Network.disable');
     expect(connectionStub.sendCommand).toHaveBeenCalledTimes(1);
     // Network still has one enable.
 
+    // @ts-ignore: Fetch is not in protocol typings yet.
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver-test.js]: can/should we update? or it's not even typed at all yet?

File: lighthouse-core/test/gather/driver-test.js (hunk 6 above)
Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js]: it was given some ungodly mega inferred type. Fixed at the import level.

File: lighthouse-core/test/gather/driver-test.js (hunk 7 above)
Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js]: ya

File: lighthouse-core/test/gather/driver-test.js (hunk 9 above)
Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js]: was gonna defer to #9459 but we can just update now

File: lighthouse-core/test/gather/driver-test.js [hunk 10]
```diff
@@ -166,8 +180,9 @@ describe('.getObjectProperty', () => {
 
 describe('.getRequestContent', () => {
   it('throws if getRequestContent takes too long', async () => {
+    // @ts-ignore
     connectionStub.sendCommand = jest.fn()
-      .mockImplementationOnce(() => new Promise(r => setTimeout(r), 5000));
+      .mockImplementationOnce(() => new Promise(r => setTimeout(r, 5000)));
```
Comment (connorjclark) [inline lighthouse-core/test/gather/driver-test.js]: this is why types are important

Comment (patrickhulce) [reply lighthouse-core/test/gather/driver-test.js]: I'm now very concerned how this was working before, wouldn't this have resolved immediately and we shouldn't have gotten protocol_timeout? 🤔

Comment (connorjclark) [reply lighthouse-core/test/gather/driver-test.js]: https://github.com/GoogleChrome/lighthouse/pull/10135
tldr: microtasks

**Comment by connorjclark** [comment]
@patrickhulce anything else you'd like to see changed here?

File: lighthouse-core/test/gather/mock-commands.js [hunk 11]
```diff
@@ -71,15 +117,25 @@ function createMockOnceFn() {
...
-    mockEvents.push({event, response});
-    return mockFn;
-  };
-
-  mockFn.findListener = event => {
-    expect(mockFn).toHaveBeenCalledWith(event, expect.anything());
-    return mockFn.mock.calls.find(call => call[0] === event)[1];
-  };
+  const mockFn = Object.assign(mockFnImpl, {
+    ...mockFnImpl,
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/mock-commands.js]: why are we assigning its spread to itself?

Comment (connorjclark) [reply lighthouse-core/test/gather/mock-commands.js]: how strange
